core/blockchain/impl/cached_tree.cpp

```diff
@@ -285,26 +141,93 @@ namespace kagome::blockchain {
       parent->children.erase(it);
     }
 
-    metadata_->leaves.erase(node->block_hash);
+    leaves_.erase(node->info.hash);
     if (parent->children.empty()) {
-      metadata_->leaves.insert(parent->block_hash);
+      leaves_.insert(parent->info.hash);
     }
 
-    auto best = metadata_->best_block.lock();
-    BOOST_ASSERT(best);
-    if (node == best) {
-      metadata_->best_block = parent;
-      for (auto it = metadata_->leaves.begin();
-           it != metadata_->leaves.end();) {
+    if (node == best_) {
+      best_ = parent;
+      for (auto it = leaves_.begin(); it != leaves_.end();) {
         const auto &hash = *it++;
-        const auto leaf_node = root_->findByHash(hash);
+        const auto leaf_node = find(hash);
         if (leaf_node == nullptr) {
           // Already removed with removed subtree
-          metadata_->leaves.erase(hash);
-        } else if (metadata_->chooseBest(leaf_node)) {
+          leaves_.erase(hash);
+        } else if (chooseBest(leaf_node)) {
           break;
         }
       }
     }
   }
+
+  CachedTree::CachedTree(const primitives::BlockInfo &root)
+      : root_{std::make_shared<TreeNode>(root)},
+        best_{root_},
+        leaves_{root.hash} {}
+
+  const std::shared_ptr<TreeNode> &CachedTree::finalized() const {
+    return root_;
+  }
+
+  const std::shared_ptr<TreeNode> &CachedTree::best() const {
+    return best_;
+  }
+
+  size_t CachedTree::leafCount() const {
+    return leaves_.size();
+  }
+
+  std::vector<primitives::BlockHash> CachedTree::leafHashes() const {
```

Comment:
Recommended to return span (or c++20 range)

Reply:
No, callers want vector